.github/workflows/ci.yml

```diff
@@ -178,3 +215,16 @@ jobs:
 #      - name: Setup tmate session
 #        uses: mxschmitt/action-tmate@v3
 #        if: ${{ failure() }}
+
+  notify-failure:
+    name: Notify Slack on Failure
+    needs: [setup, build, test, lint, integration-tests]
+    runs-on: ubuntu-latest
+    if: failure() && (github.ref == 'refs/heads/main' || github.ref == 'refs/heads/develop')
+
+    steps:
+      - name: Notify Failure
+        run: |
+          curl -X POST -H "Content-type: application/json" --data "{\"text\": \"❌ CI failed on ${{ github.ref_name }}. Check the build at ${{ github.server_url }}/${{ github.repository }}/actions/runs/${{ github.run_id }}\"}" "$SLACK_WEBHOOK_URL"
```

Comment:
Use env vars instead of interpolating with `${{ }}`.
We had that requested as a feedback from a security reviewer specialized in GH workflow vulnerabilities.